internal/pkg/client/gitlab/client.go

```diff
@@ -0,0 +1,105 @@
+// -----------------------------------------------------------------------------
+// SPDX-FileCopyrightText: Copyright © 2024 bomctl a Series of LF Projects, LLC
+// SPDX-FileName: internal/pkg/client/gitlab/client.go
+// SPDX-FileType: SOURCE
+// SPDX-License-Identifier: Apache-2.0
+// -----------------------------------------------------------------------------
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -----------------------------------------------------------------------------
+
+package gitlab
+
+import (
+	"fmt"
+	"regexp"
+
+	gitlab "gitlab.com/gitlab-org/api/client-go"
+
+	"github.com/bomctl/bomctl/internal/pkg/netutil"
+)
+
+type Client struct {
+	ProjectProvider
+	BranchProvider
+	CommitProvider
+	DependencyListExporter
+	Export      *gitlab.DependencyListExport
+	GitLabToken string
+}
+
+func (*Client) Name() string {
+	return "GitLab"
+}
+
+func (*Client) RegExp() *regexp.Regexp {
+	return regexp.MustCompile(fmt.Sprintf("(?i)^%s%s%s$",
+		`(?P<scheme>https?|git|ssh):\/\/`,
+		`(?P<hostname>[^@\/?#:]+gitlab[^@\/?#:]+)(?::(?P<port>\d+))?/`,
+		`(?P<path>[^@#]+)@(?P<branch>\S+)`))
+}
+
+func (client *Client) Parse(rawURL string) *netutil.URL {
+	results := map[string]string{}
+	pattern := client.RegExp()
+	match := pattern.FindStringSubmatch(rawURL)
+
+	for idx, name := range match {
+		results[pattern.SubexpNames()[idx]] = name
+	}
+
+	// Ensure required map fields are present.
+	for _, required := range []string{"scheme", "hostname", "path", "branch"} {
+		if value, ok := results[required]; !ok || value == "" {
+			return nil
+		}
+	}
+
+	return &netutil.URL{
+		Scheme:   results["scheme"],
+		Hostname: results["hostname"],
+		Port:     results["port"],
+		Path:     results["path"],
+		Fragment: results["branch"],
+	}
+}
+
+func NewGitLabClient(sourceURL, gitLabToken string) *Client {
```

Comment:
Since `gitlab` and `client` are already part of the import path, how do you feel about renaming this function to just `New`?

Can the token be retrieved from inside this function instead of passing it in? The environment variable should still be configurable within tests.

Reply:
Actually...is this function really needed? I think all of its logic could be moved into `Parse`; the client receiver can be updated in-place since it's a pointer.

From the `New` function in `internal/pkg/client/client.go`:

```go
	for _, client := range []Client{
		&github.Client{},
		&gitlab.Client{},
		&git.Client{},
		&http.Client{},
		&oci.Client{},
	} {
		if url := client.Parse(sbomURL); url != nil {
			return client, nil
		}
```

When `client.Parse` gets called on the gitlab client, it can set everything on the client directly. Just make sure the URL string passes all the parsing checks first, then setup the client fields, etc.

Reply:
I would have done that, but since ideally a function should do only one thing (and this one already parses and returns a URL), I opted for a factory function. What are your thoughts on that? Should I merge them into parse to also initialize the client?

Reply:
What if you just rename `NewGitlabClient` to something like `func(client *Client) initClient` (call it whatever makes sense but make it private) and call it from `Parse`?

Reply:
Sorry. I said "this one" referring to `Parse` when you were talking about `NewGitLabClient` initially. 

I'm OK with renaming `NewGitLabClient` to `new` and calling it from `Parse`. What seems unusual for me is having it initialize the instance **_and_** return a URL struct, because, from my experience outside Go, when a parse method is used to initialize an object it means "parse this data into the object", but here that would mean "parse this string into a URL struct **_and_** also initialize my "Client" object.

Reply:
I went ahead and implemented the change, but I'd like to discuss the issue I outlined above. I think we can make the way we implement clients in Bomctl a bit cleaner.